enclave/crypto/openssl/init.c

```diff
@@ -25,6 +37,52 @@ static int _initialize_symcrypt_engine()
     return (result == OE_SYMCRYPT_ENGINE_SUCCESS) ? 1 : 0;
 }
 
+#if OECRYPTO_OPENSSL_VER >= 3
+static OSSL_PROVIDER* _default_prov;
+static OSSL_PROVIDER* _scossl_prov;
+
+static void _finalize_prov(void)
+{
+    if (_default_prov)
+    {
+        OSSL_PROVIDER_unload(_default_prov);
+        _default_prov = NULL;
+    }
+    if (_scossl_prov)
+    {
+        OSSL_PROVIDER_unload(_scossl_prov);
+        _scossl_prov = NULL;
+    }
+}
+
+static int _initialize_symcrypt_provider()
+{
+    int result = 1;
+
+    _default_prov = OSSL_PROVIDER_load(NULL, "default");
+
+    if (_default_prov == NULL)
+        result = 0;
```

Comment:
Can we use the early return pattern like `_initialize_rdrand_engine()` does?

Reply:
The problem with this pattern is, if I call _finalize_prov, it will unload both the default provider and the symcrypt provider. Since we need the default provider for when symcrypt is not used, it is required that atexit(_finalize_prov) be reached, because we should only unload the default provider at program exit.

Reply:
You can call atexit earlier in the function (before even calling OSSL_PROVIDER_load). Or you can call atexit under the "done:" label.